README.md
# Treefit for Python - The first software for quantitative trajectory inference

This is an implementation of
[**Treefit**](https://hayamizu-lab.github.io/treefit/) in Python.

**Treefit** is a novel data analysis toolkit that helps you perform
two types of analysis: 1) checking the goodness-of-fit of tree models
to your single-cell gene expression data; and 2) deciding which tree
best fits your data.

## Install

```bash
pip install treefit
```

## Usage

The main functions are `treefit.treefit()` and `treefit.plot()`:

```python
fit = treefit.treefit(YOUR_SINGLE_CELL_GENE_EXPRESSION_DATA)
treefit.plot(fit)
```

See https://hayamizu-lab.github.io/treefit-python/ for details.


doc/release_notes.rst
.. Copyright (C) 2020  Momoko Hayamizu <hayamizu@ism.ac.jp>
..
.. This program is free software: you can redistribute it and/or modify
.. it under the terms of the GNU Lesser General Public License as
.. published by the Free Software Foundation, either version 3 of the
.. License, or (at your option) any later version.
..
.. This program is distributed in the hope that it will be useful,
.. but WITHOUT ANY WARRANTY; without even the implied warranty of
.. MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.. GNU Lesser General Public License for more details.
..
.. You should have received a copy of the GNU Lesser General Public
.. License along with this program.  If not, see
.. <http://www.gnu.org/licenses/>.

Release notes
=============

.. toctree::
   :maxdepth: 2
   :glob:
   :reversed:

   release_notes/*


test/test_fit.py
# Copyright (C) 2020  Momoko Hayamizu <hayamizu@ism.ac.jp>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

import numpy as np
import pytest

import treefit
import treefit.fit

def test_caluculate_distance_matrix():
    values = np.array([[1, 0],
                       [0, 1],
                       [0, 0]])
    actual_distance_matrix = treefit.fit.calculate_distance_matrix(values)
    expected_distance_matrix = np.array([[0,          np.sqrt(2), 1],
                                         [np.sqrt(2), 0,          1],
                                         [1,          1,          0]])
    assert np.all(actual_distance_matrix == expected_distance_matrix)

def test_perturbate_knn():
    expression = treefit.data.generate_2d_n_arms_star_data(100, 3, 0.1)
    n_perturbations = 5
    strength = 0.2
    min_diff_mean = 0.001
    max_diff_mean = 0.02
    min_diff_variance = 0.00001
    max_diff_variance = 0.001
    perturbated_expression_list = \
        map(lambda i: treefit.fit.perturbate_knn(expression, strength),
            range(n_perturbations))
    original_distance_matrix = \
        treefit.fit.calculate_distance_matrix(expression)
    perturbated_distance_matrix_list = \
        map(treefit.fit.calculate_distance_matrix,
            perturbated_expression_list)
    diffs = \
        map(lambda perturbated_distance_matrix: \
                abs(original_distance_matrix - perturbated_distance_matrix),
            perturbated_distance_matrix_list)
    diffs = list(diffs)
    diffs = np.array(diffs)
    diff_means = np.mean(diffs, axis=(1, 2))
    diff_variances = np.var(diffs, axis=(1, 2))
    assert np.all(diff_means > min_diff_mean)
    assert np.all(diff_means < max_diff_mean)
    assert np.all(diff_variances > min_diff_variance)
    assert np.all(diff_variances < max_diff_variance)

def test_calculate_mst():
    values = np.array([[1, 0],
                       [0, 3],
                       [0, 0]])
    mst = treefit.fit.calculate_mst(values)
    assert mst.toarray() == \
        pytest.approx(np.array([[0,     0, 1.0],
                                [0,     0, 1.0],
                                [1.0, 1.0,   0]]))

def test_calculate_low_dimension_laplacian_eigenvectors():
    values = np.stack([range(1, 11), range(1, 11)],
                      axis=1)
    mst = treefit.fit.calculate_mst(values)
    eigenvectors = \
        treefit.fit.calculate_low_dimension_laplacian_eigenvectors(mst, 4)
    assert eigenvectors[0, :] == \
        pytest.approx(np.array([-0.44170765,
                                0.4253254,
                                -0.39847023,
                                -0.3618034]))

def test_calculate_canonical_correlation():
    values1 = np.stack([range(1, 11), range(1, 11)],
                       axis=1)
    mst1 = treefit.fit.calculate_mst(values1)
    eigenvectors1 = \
        treefit.fit.calculate_low_dimension_laplacian_eigenvectors(mst1, 4)
    values2 = np.array([[5, 1, 2, 5, 8, 9, 2, 8, 1, 2],
                        [9, 7, 4, 2, 8, 3, 1, 9, 4, 1]]).T
    mst2 = treefit.fit.calculate_mst(values2)
    eigenvectors2 = \
        treefit.fit.calculate_low_dimension_laplacian_eigenvectors(mst2, 4)
    canonical_correlation = \
        treefit.fit.calculate_canonical_correlation(eigenvectors1,
                                                    eigenvectors2)
    assert canonical_correlation == \
        pytest.approx(np.array([0.94948070,
                                0.76344509,
                                0.57019530,
                                0.06708014]))

def test_calculate_grassmann_distance_max_cca():
    canonical_correlation = np.array([0.94948070,
                                      0.76344509,
                                      0.57019530,
                                      0.06708014])
    distance = \
        treefit.fit.calculate_grassmann_distance_max_cca(canonical_correlation)
    assert distance == pytest.approx(0.3138254297)

def test_calculate_grassmann_distance_rms_cca():
    canonical_correlation = np.array([0.94948070,
                                      0.76344509,
                                      0.57019530,
                                      0.06708014])
    distance = \
        treefit.fit.calculate_grassmann_distance_rms_cca(canonical_correlation)
    assert distance == pytest.approx(0.7392590158)

def test_treefit_2_arms():
    star = treefit.data.generate_2d_n_arms_star_data(200, 2, 0.1)
    fit = treefit.treefit({"expression": star},
                          "tree-like")
    assert [fit.name, fit.n_principal_paths_candidates[0]] == ["tree-like", 2]

def test_treefit_3_arms():
    star = treefit.data.generate_2d_n_arms_star_data(200, 3, 0.1)
    fit = treefit.treefit({"expression": star},
                          "tree-like")
    assert [fit.name, fit.n_principal_paths_candidates[0]] == ["tree-like", 3]


test/test_treefit.py
# Copyright (C) 2020  Momoko Hayamizu <hayamizu@ism.ac.jp>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

import numpy as np
import pytest

import treefit.generate
import treefit.treefit

def test_caluculate_distance_matrix():
    values = np.array([[1, 0],
                       [0, 1],
                       [0, 0]])
    actual_distance_matrix = treefit.treefit.calculate_distance_matrix(values)
    expected_distance_matrix = np.array([[0,          np.sqrt(2), 1],
                                         [np.sqrt(2), 0,          1],
                                         [1,          1,          0]])
    assert np.all(actual_distance_matrix == expected_distance_matrix)

def test_perturbate_knn():
    expression = treefit.generate.generate_2d_n_arms_star_data(100, 3, 0.1)
    n_perturbations = 5
    strength = 0.2
    min_diff_mean = 0.001
    max_diff_mean = 0.02
    min_diff_variance = 0.00001
    max_diff_variance = 0.001
    perturbated_expression_list = \
        map(lambda i: treefit.treefit.perturbate_knn(expression, strength),
            range(n_perturbations))
    original_distance_matrix = \
        treefit.treefit.calculate_distance_matrix(expression)
    perturbated_distance_matrix_list = \
        map(treefit.treefit.calculate_distance_matrix,
            perturbated_expression_list)
    diffs = \
        map(lambda perturbated_distance_matrix: \
                abs(original_distance_matrix - perturbated_distance_matrix),
            perturbated_distance_matrix_list)
    diffs = list(diffs)
    diffs = np.array(diffs)
    diff_means = np.mean(diffs, axis=(1, 2))
    diff_variances = np.var(diffs, axis=(1, 2))
    assert np.all(diff_means > min_diff_mean)
    assert np.all(diff_means < max_diff_mean)
    assert np.all(diff_variances > min_diff_variance)
    assert np.all(diff_variances < max_diff_variance)

def test_calculate_mst():
    values = np.array([[1, 0],
                       [0, 3],
                       [0, 0]])
    mst = treefit.treefit.calculate_mst(values)
    assert mst.toarray() == \
        pytest.approx(np.array([[0,     0, 1.0],
                                [0,     0, 1.0],
                                [1.0, 1.0,   0]]))

def test_calculate_low_dimension_laplacian_eigenvectors():
    values = np.stack([range(1, 11), range(1, 11)],
                      axis=1)
    mst = treefit.treefit.calculate_mst(values)
    eigenvectors = \
        treefit.treefit.calculate_low_dimension_laplacian_eigenvectors(mst, 4)
    assert eigenvectors[0, :] == \
        pytest.approx(np.array([-0.44170765,
                                0.4253254,
                                -0.39847023,
                                -0.3618034]))

def test_calculate_canonical_correlation():
    values1 = np.stack([range(1, 11), range(1, 11)],
                       axis=1)
    mst1 = treefit.treefit.calculate_mst(values1)
    eigenvectors1 = \
        treefit.treefit.calculate_low_dimension_laplacian_eigenvectors(mst1, 4)
    values2 = np.array([[5, 1, 2, 5, 8, 9, 2, 8, 1, 2],
                        [9, 7, 4, 2, 8, 3, 1, 9, 4, 1]]).T
    mst2 = treefit.treefit.calculate_mst(values2)
    eigenvectors2 = \
        treefit.treefit.calculate_low_dimension_laplacian_eigenvectors(mst2, 4)
    canonical_correlation = \
        treefit.treefit.calculate_canonical_correlation(eigenvectors1,
                                                        eigenvectors2)
    assert canonical_correlation == \
        pytest.approx(np.array([0.94948070,
                                0.76344509,
                                0.57019530,
                                0.06708014]))

def test_calculate_grassmann_distance_max_cca():
    canonical_correlation = np.array([0.94948070,
                                      0.76344509,
                                      0.57019530,
                                      0.06708014])
    distance = \
        treefit.treefit.calculate_grassmann_distance_max_cca(canonical_correlation)
    assert distance == pytest.approx(0.3138254297)

def test_calculate_grassmann_distance_rms_cca():
    canonical_correlation = np.array([0.94948070,
                                      0.76344509,
                                      0.57019530,
                                      0.06708014])
    distance = \
        treefit.treefit.calculate_grassmann_distance_rms_cca(canonical_correlation)
    assert distance == pytest.approx(0.7392590158)

def test_treefit_2_arms():
    star = treefit.generate.generate_2d_n_arms_star_data(200, 2, 0.1)
    fit = treefit.treefit.treefit({"expression": star},
                                  "tree-like")
    assert [fit.name, fit.n_principal_paths_candidates[0]] == ["tree-like", 2]

def test_treefit_3_arms():
    star = treefit.generate.generate_2d_n_arms_star_data(200, 3, 0.1)
    fit = treefit.treefit.treefit({"expression": star},
                                  "tree-like")
    assert [fit.name, fit.n_principal_paths_candidates[0]] == ["tree-like", 3]


treefit/data.py
# Copyright (C) 2020  Momoko Hayamizu <hayamizu@ism.ac.jp>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

import numpy as np

def generate_2d_n_arms_star_data(n_samples, n_arms, fatness):
    """Generate a 2-dimensional star tree data that contain ``n_samples``
    data points and fit a star tree with ``n_arms`` arms.

    Parameters
    ----------
    n_samples : int
        The number of samples to be generated.

    n_arms : int
        The number of arms to be generated.

    fatness : float
        How fat from the based star tree. ``[0.0, 1.0]`` is available
        value range.

    Returns
    -------
    star : numpy.array

        A generated ``numpy.array``. The rows and columns correspond
        to samples and features.

    Examples
    --------
    >>> import treefit
    >>> from matplotlib.pyplot as plt
    # Generate a 2-dimensional star tree data that contain 500 data points
    # and fit a star tree with 3 arms. The generated data are a bit noisy but
    # tree-like.
    >>> star_tree_like = treefit.data.generate_2d_n_arms_star_data(500, 3, 0.1)
    >>> plt.figure()
    >>> plt.scatter(star_tree_like[:, 0], star_tree_like[:, 1])
    # Generate a 2-dimensional star tree data that contain 600 data points
    # and fit a star tree with 5 arms. The generated data are very noisy and
    # less tree-like.
    >>> star_less_tree_like = treefit.data.generate_2d_n_arms_star_data(600, 5, 0.9)
    >>> plt.figure()
    >>> plt.scatter(star_less_tree_like[:, 0], \
    ...             star_less_tree_like[:, 1])
    """
    n_features = 2
    standard_deviation = fatness / n_arms
    star = np.zeros((n_samples, n_features), np.float)
    for i in range(n_samples):
        arm = np.random.choice(range(n_arms))
        theta = (arm + 1) / n_arms * n_features * np.pi
        position = np.array([np.cos(theta), np.sin(theta)])
        position = position * np.random.uniform()
        position = position + np.random.normal(scale=standard_deviation,
                                               size=n_features)
        star[i, :] = position
    return star

def generate_2d_n_arms_linked_star_data(n_samples_list,
                                        n_arms_list,
                                        fatness):
    """Generate a 2-dimensional linked star tree data.

    Each star tree data contain ``n_samples_vector[i]`` data points and
    fit a star tree with ``n_arms_vector[i]`` arms.

    Parameters
    ----------
    n_samples_list : [int]
        The list of the number of samples to be generated. For
        example, ``[200, 100, 300]`` means that the first tree has 200
        samples, the second tree has 100 samples and the third tree
        has 300 samples.

    n_arms_list : [int]
        The list of the number of arms to be generated. For example,
        ``[3, 2, 5]`` means the first tree fits a star tree with 3
        arms, the second tree fits a star tree with 2 arms and the
        third tree fits a star tree with 5 arms. The length of
        ``n_arms_list`` must equal to the length of
        ``n_samples_list``.

    fatness : [float]
        How fat from the based tree. ``[0.0, 1.0]`` is available value
        range.

    Returns
    -------
    linked_star : numpy.array

        A generated `numpy.array`. The rows and columns correspond to
        samples and features.

    Examples
    --------
    >>> import treefit
    >>> from matplotlib.pyplot as plt
    # Generate a 2-dimensional linked star tree data that contain
    # 200-400-300 data points and fit a linked star tree with 3-5-4
    # arms. The generated data are a bit noisy but tree-like.
    >>> linked_star_tree_like = \
    ...     treefit.data.generate_2d_n_arms_linked_star_data([200, 400, 300],
    ...                                                      [3, 5, 4],
    ...                                                      0.1)
    >>> plt.figure()
    >>> plt.scatter(linked_star_tree_like[:, 0],
    ...             linked_star_tree_like[:, 1])
    # Generate a 2-dimensional linked star tree data that contain
    # 300-200 data points and fit a linked star tree with 4-3 arms.
    # The generated data are very noisy and less tree-like.
    >>> linked_star_less_tree_like = \
    ...     treefit.data.generate_2d_n_arms_linked_star_data([300, 200],
    ...                                                      [4, 3],
    ...                                                      0.9)
    >>> plt.figure()
    >>> plt.scatter(linked_star_less_tree_like[:, 0],
    ...             linked_star_less_tree_like[:, 1])
    """

    n_features = 2
    n_total_samples = np.sum(n_samples_list)
    star = np.zeros((n_total_samples, n_features), np.float)
    n_samples_offset = 0
    sub_star_offsets = [0.0, 0.0]
    for i in range(len(n_samples_list)):
        n_samples = n_samples_list[i]
        n_arms = n_arms_list[i]
        sub_star = generate_2d_n_arms_star_data(n_samples, n_arms, fatness)
        theta = 2 * np.pi * (n_arms // 2 / n_arms)
        sub_star_offsets[0] = sub_star_offsets[0] + -np.cos(theta) + 1
        sub_star_offsets[1] = sub_star_offsets[1] + -np.sin(theta)
        sub_star[:, 0] = sub_star[:, 0] + sub_star_offsets[0]
        sub_star[:, 1] = sub_star[:, 1] + sub_star_offsets[1]
        star[n_samples_offset:(n_samples_offset+n_samples), :] = sub_star
        n_samples_offset += n_samples
    return star

def generate_2d_n_arms_star_data(n_samples, n_arms, fatness):
    """Generate a 2-dimensional star tree data that contain ``n_samples``
    data points and fit a star tree with ``n_arms`` arms.

    Parameters
    ----------
    n_samples : int
        The number of samples to be generated.

    n_arms : int
        The number of arms to be generated.

    fatness : float
        How fat from the based star tree. ``[0.0, 1.0]`` is available
        value range.

    Returns
    -------
    star : numpy.array

        A generated ``numpy.array``. The rows and columns correspond
        to samples and features.

    Examples
    --------
    >>> import treefit
    >>> from matplotlib.pyplot as plt
    # Generate a 2-dimensional star tree data that contain 500 data points
    # and fit a star tree with 3 arms. The generated data are a bit noisy but
    # tree-like.
    >>> star_tree_like = treefit.data.generate_2d_n_arms_star_data(500, 3, 0.1)
    >>> plt.figure()
    >>> plt.scatter(star_tree_like[:, 0], star_tree_like[:, 1])
    # Generate a 2-dimensional star tree data that contain 600 data points
    # and fit a star tree with 5 arms. The generated data are very noisy and
    # less tree-like.
    >>> star_less_tree_like = treefit.data.generate_2d_n_arms_star_data(600, 5, 0.9)
    >>> plt.figure()
    >>> plt.scatter(star_less_tree_like[:, 0], \
    ...             star_less_tree_like[:, 1])
    """
    n_features = 2
    standard_deviation = fatness / n_arms
    star = np.zeros((n_samples, n_features), np.float)
    for i in range(n_samples):
        arm = np.random.choice(range(n_arms))
        theta = (arm + 1) / n_arms * n_features * np.pi
        position = np.array([np.cos(theta), np.sin(theta)])
        position = position * np.random.uniform()
        position = position + np.random.normal(scale=standard_deviation,
                                               size=n_features)
        star[i, :] = position
    return star


treefit/__init__.py
# Copyright (C) 2020  Momoko Hayamizu <hayamizu@ism.ac.jp>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

__version__ = '1.0.1'

import treefit.data
import treefit.fit

"""Shortcut"""
plot = treefit.fit.plot

"""Shortcut"""
treefit = treefit.fit.treefit


treefit/fit.py
# Copyright (C) 2020  Momoko Hayamizu <hayamizu@ism.ac.jp>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

import math
import matplotlib.pyplot as plt
import matplotlib.ticker
import numpy as np
import pandas as pd
import scipy
import sklearn.decomposition

def perturbate_poisson(counts, strength=1.0):
    return (np.random.poisson(counts) * strength).astype(counts.dtype)

def calculate_distance_matrix(expression):
    # Use 'euclidean' method for now.
    # TODO: make method parameter
    return scipy.spatial.distance.cdist(expression, expression, 'euclidean')

def calculate_mst(expression):
    distance_matrix = calculate_distance_matrix(expression)
    mst = scipy.sparse.csgraph.minimum_spanning_tree(distance_matrix)
    # Remove weights
    mst[mst > 0] = 1
    return mst + mst.T

def perturbate_knn(expression, strength=1.0):
    n_samples, n_features = expression.shape
    # TODO: Improve
    k_nearest_neighbors = round(n_samples * 0.0125)
    if k_nearest_neighbors < 2:
        k_nearest_neighbors = 2
    standard_deviation = strength / np.sqrt(k_nearest_neighbors)
    distance_matrix = calculate_distance_matrix(expression)
    perturbated_expression = np.zeros((n_samples, n_features))
    for i in range(n_samples):
        sorted_indices = np.argsort(distance_matrix[i, :])
        nearest_neighbors = \
            expression[sorted_indices[1:(k_nearest_neighbors + 1)]]
        diffs = nearest_neighbors - expression[i]
        weights = np.random.normal(scale=standard_deviation,
                                   size=(n_features,))
        weighted_diffs = diffs * weights
        perturbated_expression[i] = expression[i] + np.sum(weighted_diffs, axis=0)
    return perturbated_expression

def calculate_low_dimension_laplacian_eigenvectors(mst, p):
    laplacian = scipy.sparse.csgraph.laplacian(mst)
    eigenvalues, eigenvectors = np.linalg.eigh(laplacian.toarray())
    while len(eigenvalues) > 0 and math.isclose(eigenvalues[0], 0, abs_tol=1e-9):
        eigenvalues = eigenvalues[1:]
        eigenvectors = eigenvectors[:, 1:]
    low_dimension_values = eigenvalues[0:p]
    low_dimension_vectors = eigenvectors[:, 0:p]
    if len(np.unique(low_dimension_values)) != len(low_dimension_values):
        low_dimension_vectors = scipy.linalg.orth(low_dimension_vectors)
    low_dimension_vectors /= np.linalg.norm(low_dimension_vectors, axis=0)
    return low_dimension_vectors

def calculate_canonical_correlation(u, v):
    uTv = np.matmul(u.T, v)
    return scipy.linalg.svd(uTv)[1]

def calculate_grassmann_distance_max_cca(canonical_correlation):
    max_cos_theta = np.max(canonical_correlation)
    return np.sqrt(np.max([0, 1 - max_cos_theta ** 2]))

def calculate_grassmann_distance_rms_cca(canonical_correlation):
    return np.sqrt(np.mean(1 - canonical_correlation ** 2))

def calculate_eigenvectors_list(original,
                                perturbations,
                                normalize,
                                reduce_dimension,
                                build_tree,
                                max_p,
                                n_perturbations):
    poisson_strength = 1.0
    # TODO: Improve
    knn_strength = 0.2 * (500 / 200) ** 0.5
    targets = []

    if perturbations is None or 'poisson' in perturbations:
        counts = original.get('counts')
        if counts:
            targets.append(counts)
            for i in range(n_perturbations):
                perturbated_counts = perturbate_poisson(counts, poisson_strength)
                targets.append(perturbated_counts)
        elif not perturbations is None:
            raise TypeError('no count data: %s' % original)

        if len(targets) > 0:
            # TODO: Normalize
            pass

    if perturbations is None or 'knn' in perturbations:
        if len(targets) == 0:
            expression = original.get('expression')
            if expression is None:
                raise TypeError('no expression data: %s' % original)

            targets.append(expression)
            for i in range(n_perturbations):
                perturbated_expression = perturbate_knn(expression, knn_strength)
                targets.append(perturbated_expression)
        else:
            targets = \
                [perturbate_knn(expression, knn_strength) for target in targets]

    def calculate(target):
        if reduce_dimension is None or isinstance(reduce_dimension, int):
            if isinstance(reduce_dimension, int):
                n_dimensions = reduce_dimension
            else:
                n_dimensions = None
            pca = sklearn.decomposition.PCA(n_dimensions)
            target = pca.fit_transform(target)
        elif reduce_dimension:
            target = reduce_dimension(target)
        tree = calculate_mst(target)
        return calculate_low_dimension_laplacian_eigenvectors(tree, max_p)

    return list(map(calculate, targets))

class Fit:
    """The estimated result of :py:func:`treefit.treefit`.

    Attributes
    ----------

    max_cca_distance: pandas.DataFrame
        The result of max canonical correlation analysis distance.

        It has the following columns:

        * ``p``: Dimensionality of the feature space of tree
          structures.

        * ``mean``: The mean of the target distance values.

        * ``standard_deviation``: The standard deviation of the target
          distance values.

    rms_cca_distance: pandas.DataFrame
        The result of root mean square canonical correlation analysis
        distance.

        This has the same columns as ``max_cca_distance``.

    n_principal_paths_candidates: [int]
        The candidates of the number of principal paths.
    """

    def __init__(self,
                 name,
                 max_cca_distance,
                 rms_cca_distance,
                 n_principal_paths_candidates):
        self.name = name
        self.max_cca_distance = max_cca_distance
        self.rms_cca_distance = rms_cca_distance
        self.n_principal_paths_candidates = n_principal_paths_candidates

    def __str__(self):
        class_name = f'{self.__class__.__module__}.{self.__class__.__qualname__}'
        return f"""{class_name}: {self.name}
max_cca_distance:
{self.max_cca_distance}
rms_cca_distance:
{self.rms_cca_distance}
n_principal_paths_candidates:
{self.n_principal_paths_candidates}"""

def treefit(target,
            name=None,
            perturbations=None,
            normalize=None,
            reduce_dimension=None,
            build_tree=None,
            max_p=20,
            verbose=False,
            n_perturbations=20):
    """Estimate the goodness-of-fit between tree models and data.

    Parameters
    ----------
    target : dict
        The target data to be estimated. It must be one of them:

        * ``{"counts": COUNTS}``
        * ``{"expression": EXPRESSION}``

        ``COUNTS`` and ``EXPRESSION`` are ``numpy.array``. The rows
        and columns correspond to samples such as cells and features
        such as genes. ``COUNTS``'s value is count data such as the
        number of genes expressed. ``EXPRESSION``'s value is
        normalized count data.

    name : string
        The name of target as string.

    perturbations : list
        How to perturbate the target data.

        If this is ``None``, all available perturbation methods are
        used.

        You can specify used perturbation methods as ``list``. Here
        are available methods:

        * ``"poisson"``: A perturbation method for counts data.
        * ``"knn"``: A perturbation method for expression data.

    normalize : callable
        How to normalize counts data.

        If this is ``None``, the default normalization is applied.

        You can specify a ``callable`` object that normalized counts
        data.

    reduce_dimension : callable
        How to reduce dimension of normalized count data.

        If this is ``None``, the default dimensionality reduction is
        applied.

        You can specify a ``callable`` object that reduces dimension
        of normalized counts data.

    build_tree : callable
        How to build a tree of expression data.

        If this is ``None``, MST is built.

        You can specify a function that builds tree of normalized
        counts data.

    max_p : int
        How many low dimension Laplacian eigenvectors are used.

        The default is ``20``.

    n_perturbations : int
        How many times to perturb.

        The default is `20`.

    Returns
    -------
    fit : treefit.fit.Fit

        An estimated result as a :py:class:`treefit.fit.Fit` object.

    Examples
    --------

    >>> import treefit
    # Generate a star tree data that have normalized expression values
    # not count data.
    >>> star = treefit.data.generate_2d_n_arms_star_data(300, 3, 0.1)
    # Estimate tree-likeness of the tree data.
    >>> fit = treefit.treefit({"expression": star})
    """

    if name is None:
        name = "fit"
    eigenvectors_list = calculate_eigenvectors_list(target,
                                                    perturbations,
                                                    normalize,
                                                    reduce_dimension,
                                                    build_tree,
                                                    max_p,
                                                    n_perturbations)
    ps = []
    max_cca_distance_means = []
    max_cca_distance_standard_deviations = []
    rms_cca_distance_means = []
    rms_cca_distance_standard_deviations = []
    for p in range(1, max_p + 1):
        ps.append(p)
        max_cca_distance_values = []
        rms_cca_distance_values = []
        for i in range(1, len(eigenvectors_list)):
            u = eigenvectors_list[0][:, 0:p]
            v = eigenvectors_list[i][:, 0:p]
            canonical_correlation = calculate_canonical_correlation(u, v)
            max_cca_distance_values.append(
                calculate_grassmann_distance_max_cca(canonical_correlation))
            rms_cca_distance_values.append(
                calculate_grassmann_distance_rms_cca(canonical_correlation))
        max_cca_distance_means.append(np.mean(max_cca_distance_values))
        max_cca_distance_standard_deviations.append(
            np.std(max_cca_distance_values))
        rms_cca_distance_means.append(np.mean(rms_cca_distance_values))
        rms_cca_distance_standard_deviations.append(
            np.std(rms_cca_distance_values))

    n_principal_paths_candidates = []
    for p in range(1, max_p - 1):
        if p == 1:
            rms_cca_distance_mean_before = float("inf")
        else:
            rms_cca_distance_mean_before = rms_cca_distance_means[p - 2]
        rms_cca_distance_mean = rms_cca_distance_means[p - 1]
        rms_cca_distance_mean_after = rms_cca_distance_means[p]
        if rms_cca_distance_mean_before > rms_cca_distance_mean and \
           rms_cca_distance_mean < rms_cca_distance_mean_after:
            n_principal_paths_candidates.append(p + 1)

    max_cca_distance = pd.DataFrame({
        'p': ps,
        'mean': max_cca_distance_means,
        'standard_deviation': max_cca_distance_standard_deviations,
    })
    rms_cca_distance = pd.DataFrame({
        'p': ps,
        'mean': rms_cca_distance_means,
        'standard_deviation': rms_cca_distance_standard_deviations,
    })
    return Fit(name,
               max_cca_distance,
               rms_cca_distance,
               n_principal_paths_candidates)

def plot(*fits):
    """Plot estimated results to get insight.

    Parameters
    ----------
    *fits : [treefit.fit.Fit]
        The estimated results by treefit.treefit() to be visualized.

    Examples
    --------
    >>> import treefit
    # Generate a tree data.
    >>> tree = treefit.data.generate_2d_n_arms_star_data(200, 3, 0.1)
    # Estimate the goodness-of-fit between tree models and the tree data.
    >>> fit = treefit.treefit({"expression": tree}, "tree")
    # Visualize the estimated result.
    >>> treefit.plot(fit)
    # You can mix multiple estimated results by adding "name" column.
    >>> tree2 = treefit.data.generate_2d_n_arms_star_data(200, 3, 0.9)
    >>> fit2 = treefit.treefit({"expression": tree2}, "tree2")
    >>> treefit.plot(fit, fit2)
    """

    fig, axes = plt.subplots(1, 2, figsize=(10, 6))
    max_ax = axes[0]
    rms_ax = axes[1]

    def plot_data_frame(ax, title, value_label, data_frame):
        p = data_frame['p']
        mean = data_frame['mean']
        standard_deviation = data_frame['standard_deviation']
        ax.set_title(title)
        ax.set_xlabel('p: Dimensionality of the feature space of trees')
        ax.set_ylabel('%s (mean and SD)' % value_label)
        ax.plot(p, mean)
        ax.fill_between(p,
                        mean - standard_deviation,
                        mean + standard_deviation,
                        alpha=0.2,
                        zorder=-10)
        ax.xaxis.set_major_locator(matplotlib.ticker.MultipleLocator(1))

    for fit in fits:
        plot_data_frame(max_ax,
                        'Analysis of the structural instability\n' +
                        'of the estimated trees',
                        'max_cca_distance',
                        fit.max_cca_distance)
        plot_data_frame(rms_ax,
                        'Prediction for\nthe number of principal paths',
                        'rms_cca_distance',
                        fit.rms_cca_distance)

    if len(fits) > 1:
        legend = [fit.name for fit in fits]
        max_ax.legend(legend)
        rms_ax.legend(legend)
